keps/sig-api-machinery/1630-kubeconfig-qps-burst/README.md

```diff
@@ -0,0 +1,130 @@
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Kubernetes REST clients use a simple token bucket rate limiter to control the pace of requests from a client.
+We will expose the values already present in the [REST config](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/rest/config.go#L114-L120), via a kubeconfig file inside the client stanza.
+We will wire the in-cluster kubeconfig file to honor the presence of a new file to allow customization inside a cluster at the discretion of a different, user provided, controller.
+
+## Motivation
+
+We've always had reasons to override these values, so we created one-off arguments (`--kube-api-burst`, `--kube-api-qps`).
+With widespread adoption of operators and other in-cluster workloads, having a standard way to provide these values makes sense.
+In addition, with the introduction of API priority and fairness, being able to configure and even disable this client-side rate limiting in compatible environments is important.
+
+### Goals
+
+1. Allow easy and consistent configuration for the existing client-side rate limiting.
+
+### Non-Goals
+
+1. Allow customization of client-side rate limiting strategies.
+
+## Proposal
+
+Add to the `AuthInfo` section of a kubeconfig.
+This section identifies a particular client identity.
+This identity spans all the contexts with different namespaces, so it is a closer logical match.
+You don't generally want different QPS characteristics based on namespace.
+
+```go
+type AuthInfo struct {
+	// ...
+	// QPS indicates the maximum QPS to the master from this client.
+	// If it's zero, the created RESTClient will use DefaultQPS: 5
+	QPS float32
+
+	// Maximum burst for throttle.
+	// If it's zero, the created RESTClient will use DefaultBurst: 10.
+	Burst int32
+}
+```
+
+### in-cluster kubeconfig overrides
+The in-cluster config is the automatic configuration that exists within pods.
+The client code itself can be written to consume an overrides.kubeconfig file, which will be merged with the standard
+kubeconfig merging rules that we use the env var path today.
+The controller code will *not* be updated.
+If a cluster-admin wants to control this file, he can provide his own controller to write content into this file in 
+serviceaccount token secrets.
+
+### Test Plan
+
+This is heavily unit testable.
+
+### Graduation Criteria
+
+The API itself is well-known since near the beginning of kube.
+We are simply exposing existing values, it will enter as GA.
+
+### Upgrade / Downgrade Strategy
+
+Old clients will ignore the unknown field.
+Old `kubectl` will rewrite the stanza and remove the new values.
+
+### Version Skew Strategy
+
+Old clients will ignore the unknown field.
+Old `kubectl` will rewrite the stanza and remove the new values.
+
+## Implementation History
+
+## Drawbacks
+
+## Alternatives
```

Comment:
Another alternative might be to plumb these via env var. That'd be gross, but very direct.

Reply:
Yeah.

Reply:
The go team tends to do things this way, e.g. for disabling their home grown DNS resolver, and also for disabling HTTP2 I think.

The benefit of this is that it only requires one component to change, no API is baked in, we can remove the env var without breaking people who are setting it.

The downside is that it's not very discoverable (but maybe that's OK since we mostly want to use it to run tests).

Reply:
Anyone setting this should also look at the client's metrics to verify that it's not spending any time blocked on the QPS limiter (I think there's already a metric).